### ISSUE TITLE
Helpdesk: Unknown installation issue

### ISSUE THREAD

**Issue by Leopere** [rr-/szurubooru#287]
I realize this is title gore but just saying your docker implementation is really rough.  Theres a fella over on https://old.reddit.com/r/selfhosted/comments/dl8mtf/booru_nightmares/ banging his head against the wall trying to figure out how to get a booru going.  I don't really need this but I tried getting a git clone-able option for the guy but stopped after reaching this far.

`docker-compose.yml`
```YAML
## Example Docker Compose configuration
##
## Use this as a template to set up docker-compose, or as guide to set up other
## orchestration services
version: '2'

services:
  sql:
    image: postgres:11-alpine
    restart: always
    environment:
      POSTGRES_USER:
      POSTGRES_PASSWORD:
    volumes:
      - "${MOUNT_SQL}:/var/lib/postgresql/data"

  elasticsearch:
    image: docker.elastic.co/elasticsearch/elasticsearch-oss:6.3.1
    restart: always
    environment:
      ## Specifies the Java heap size used
      ## Read
      ##  https://www.elastic.co/guide/en/elasticsearch/reference/current/docker.html
      ## for more info
      ES_JAVA_OPTS: -Xms512m -Xmx512m
    volumes:
      - index:/usr/share/elasticsearch/data
    ulimits:
      nofile:
        soft: 65536
        hard: 65536

  server:
    image: rrdash/szurubooru-server:latest
    restart: always
    depends_on:
      - sql
      - elasticsearch
    environment:
      ## These should be the names of the dependent containers listed above,
      ## or FQDNs/IP addresses if these services are running outside of Docker
      POSTGRES_HOST: sql
      ESEARCH_HOST: elasticsearch
      ## Credentials for database:
      POSTGRES_USER:
      POSTGRES_PASSWORD:
      ## Commented Values are Default:
      #POSTGRES_DB: defaults to same as POSTGRES_USER
      #POSTGRES_PORT: 5432
      #ESEARCH_PORT: 9200
      #ESEARCH_INDEX: szurubooru
      #ESEARCH_PASSWORD: (empty by default, set if you are using an external
      #                  source for elasticsearch and want to use HTTP basic
      #                  authentication for security)
      #ESEARCH_USER: szurubooru (only used if password is set)
      #LOG_SQL: 0 (1 for verbose SQL logs)
    volumes:
      - "${MOUNT_DATA}:/data"
      - "./config.yaml:/opt/app/config.yaml"

  client:
    image: rrdash/szurubooru-client:latest
    restart: always
    depends_on:
      - server
    environment:
      BACKEND_HOST: server
      BASE_URL: ${BASE_URL}
    volumes:
      - "${MOUNT_DATA}:/data:ro"
    ports:
      - "${PORT}:80"

volumes:
  index: # Scratch space for ElasticSearch index, will be rebuilt if lost
```
`config.yaml`
```YAML
# rather than editing this file, it is strongly suggested to create config.yaml
# and override only what you need.

# shown in the website title and on the front page
name: szurubooru
# full url to the homepage of this szurubooru site, with no trailing slash
domain: # example: http://example.com
# user agent name used to download files from the web on behalf of the api users
user_agent:
# used to salt the users' password hashes and generate filenames for static content
secret: changeme

# Delete thumbnails and source files on post delete
# Original functionality is no, to mitigate the impacts of admins going
# on unchecked post purges.
delete_source_files: no

thumbnails:
    avatar_width: 300
    avatar_height: 300
    post_width: 300
    post_height: 300

convert:
  gif:
    to_webm: false
    to_mp4: false

# used to send password reset e-mails
smtp:
    host: # example: localhost
    port: # example: 25
    user: # example: bot
    pass: # example: groovy123
    from: # example: noreply@example.com
    # if host is left empty the password reset feature will be disabled, in which case it is
    # recommended to fill contactEmail so that users know who to contact when they want to reset their password

contact_email: # example: bob@example.com. Meant for manual password reset procedures

enable_safety: yes

tag_name_regex: ^\S+$
tag_category_name_regex: ^[^\s%+#/]+$

# don't make these more restrictive unless you want to annoy people; if you do
# customize them, make sure to update the instructions in the registration form
# template as well.
password_regex: '^.{5,}$'
user_name_regex: '^[a-zA-Z0-9_-]{1,32}$'

default_rank: regular

privileges:
    'users:create:self':            anonymous # Registration permission
    'users:create:any':             administrator
    'users:list':                   regular
    'users:view':                   regular
    'users:edit:any:name':          moderator
    'users:edit:any:pass':          moderator
    'users:edit:any:email':         moderator
    'users:edit:any:avatar':        moderator
    'users:edit:any:rank':          moderator
    'users:edit:self:name':         regular
    'users:edit:self:pass':         regular
    'users:edit:self:email':        regular
    'users:edit:self:avatar':       regular
    'users:edit:self:rank':         moderator # one can't promote themselves or anyone to upper rank than their own.
    'users:delete:any':             administrator
    'users:delete:self':            regular

    'user_tokens:list:any':         administrator
    'user_tokens:list:self':        regular
    'user_tokens:create:any':       administrator
    'user_tokens:create:self':      regular
    'user_tokens:edit:any':         administrator
    'user_tokens:edit:self':        regular
    'user_tokens:delete:any':       administrator
    'user_tokens:delete:self':      regular

    'posts:create:anonymous':       regular
    'posts:create:identified':      regular
    'posts:list':                   anonymous
    'posts:reverse_search':         regular
    'posts:view':                   anonymous
    'posts:view:featured':          anonymous
    'posts:edit:content':           power
    'posts:edit:flags':             regular
    'posts:edit:notes':             regular
    'posts:edit:relations':         regular
    'posts:edit:safety':            power
    'posts:edit:source':            regular
    'posts:edit:tags':              regular
    'posts:edit:thumbnail':         power
    'posts:feature':                moderator
    'posts:delete':                 moderator
    'posts:score':                  regular
    'posts:merge':                  moderator
    'posts:favorite':               regular
    'posts:bulk-edit:tags':         power
    'posts:bulk-edit:safety':       power

    'tags:create':                  regular
    'tags:edit:names':              power
    'tags:edit:category':           power
    'tags:edit:description':        power
    'tags:edit:implications':       power
    'tags:edit:suggestions':        power
    'tags:list':                    regular
    'tags:view':                    anonymous
    'tags:merge':                   moderator
    'tags:delete':                  moderator

    'tag_categories:create':        moderator
    'tag_categories:edit:name':     moderator
    'tag_categories:edit:color':    moderator
    'tag_categories:list':          anonymous
    'tag_categories:view':          anonymous
    'tag_categories:delete':        moderator
    'tag_categories:set_default':   moderator

    'comments:create':              regular
    'comments:delete:any':          moderator
    'comments:delete:own':          regular
    'comments:edit:any':            moderator
    'comments:edit:own':            regular
    'comments:list':                regular
    'comments:view':                regular
    'comments:score':               regular

    'snapshots:list':               power

    'uploads:create':               regular

## ONLY SET THESE IF DEPLOYING OUTSIDE OF DOCKER
#debug: 0 # generate server logs?
#show_sql: 0 # show sql in server logs?
#data_url: /data/
#data_dir: /var/www/data
## usage: schema://user:password@host:port/database_name
## example: postgres://szuru:dog@localhost:5432/szuru_test
#database:
#elasticsearch: # used for reverse image search
#    host: localhost
#    port: 9200
#    index: szurubooru
#    user: szurubooru
#    pass:
```
`.env`
```BASH
# Database credentials
POSTGRES_USER=szuru
POSTGRES_PASSWORD=changeme

# This shows up on the homescreen, indicating build information
BUILD_INFO=latest

# Port to expose HTTP service
# Set to 127.0.0.1:8080 if you wish to reverse-proxy the docker's port,
# otherwise the port specified here will be publicly accessible
PORT=8080

# URL base to run szurubooru under
# See "Additional Features" section in INSTALL.md
BASE_URL=/

# Directory to store image data
MOUNT_DATA=./data

# Directory to store database files
MOUNT_SQL=./sql
```

**Comment by sgsunder** [comment]
Ok... so you just linked me a rant post on reddit, provided no details on the errors you encountered (logfiles, stack traces, etc), provided no details on the troubleshooting steps you attempted, and expect me to do what with this exactly?

**Comment by Leopere** [comment]
I guess be magic.  I gave you the docker-compose.yml file, I gave you the config I linked you the reason I'm here, and I mentioned that it's not working.  I realize "it doesn't work" isn't exactly a good bug report but as far as I can tell there's no clear path to success given the install.md but if you wish to rename the git issue to something less useful then that's fine.

I followed the guide relatively closely I just pulled the files that I included in my post out of what you would've had me `git clone` your entire repository for and put them all three in the same directory and adjusted the volume mounts and I keep getting Elastic Search cycling repeatedly failing to launch. 

I figured that ElasticSearch was failing because I'd tried mounting it directly to the filesystem as all of the other mounts are directly to a directory relative to the `docker-compose.yml`. So I switched back to your version of the file mount using a docker volume and still got more errors in Elastic Search using your near default docker-compose.yml of which I'd included the entire configuration as I ran it locally for you here in this thread.  I figured posting the entire environment as I was running it here would've been sufficient but I guess you need more.  

If you have an article to differ me to on where to get enough information to successfully interface with your GitHub issues I wouldn't mind reading that.

**Comment by sgsunder** [comment]
I realize that dealing with installation issues can be really frustrating.

However, that does not excuse you to vent your frustrations onto the maintainers of a small, open-source project such as this. Maintaining Szurubooru is one of many side-projects for me, alongside my full-time job as a mechanical engineer (read: not a full-stack developer), and various other responsibilities.

If you have an issue with the documentation, you could have put in a pull request suggesting an improvement, instead of complaining to the maintainer about how much the documentation sucks. If you look through the issue tracker history, you'll see that I make an effort to review pull requests, address minor bugs that pop up, and help users with issues they encounter, when they present it in a polite manner.

> I gave you the docker-compose.yml file, I gave you the config I linked you the reason I'm here, and I mentioned that it's not working.

You could at least highlight what differences you made from the default config files instead of providing me with a near-verbatim dump of them. As far as I can tell, you only seemed to have changed the volume mount points correct? If so, why waste my time and have me read the entirety of the config?

> I just pulled the files that I included in my post out of what you would've had me `git clone` your entire repository for

We very recently switched over to Docker Hub hosted builds and are still working out the kinks in it. I haven't gotten around to making the minor edits to the install instructions to account for this fact. But this attitude isn't doing you any favors, either. Pretty much the only change that would need to be made is dropping the `docker-compose build` command, and even leaving it in there doesn't cause any errors (The build simply gets skipped). So it really just sounds like you're being petty for the sake of it.

> I figured that ElasticSearch was failing because I'd tried mounting it directly to the filesystem as all of the other mounts are directly to a directory relative to the docker-compose.yml. So I switched back to your version of the file mount using a docker volume and still got more errors in Elastic Search using your near default docker-compose.yml of which I'd included the entire configuration as I ran it locally for you here in this thread. I figured posting the entire environment as I was running it here would've been sufficient but I guess you need more.

I copy-pasted the configuration dump you gave me onto my system, and it started up just fine, no issues with elasticsearch or uploading files on my end. Therefore, I will repeat: **Unless you post a stacktrace or logfile, I have no idea on how to even begin to help you debug the issue that you're facing.**

Furthermore, if you have complaints specifically about the Elasticsearch docker container, I would encourage you to take to the [Elasticsearch project](https://github.com/elastic/elasticsearch/issues) instead. Unlike me, they are actually a professional dev team, and can help you better.

**Comment by noirscape** [comment]
Elasticsearch failing is probably caused by not raising the `vm.max_map_count` parameter.

Installation tells you to raise it before starting the installation, linking to [this page](https://www.elastic.co/guide/en/elasticsearch/reference/current/docker.html#docker-cli-run-prod-mode).

**Comment by sgsunder** [comment]
Closing issue as I'm unable to reproduce.

**Comment by Leopere** [comment]
@noirscape that's useful thanks for the reply it will be useful if anyone else is having this trouble getting this running in the future and stumbles across it.

**Comment by Leopere** [comment]
@sgsunder I swore that I posted more context to the issue I was passing through on /r/selfhosted and figured I'd offer a bit of assistance to the guy on Reddit.  I thought that figuring out how to running Szurubooru might be a fun exercise, and when I hit a wall, I figured I'd raise it with you.  This entire miscommunication is my fault; I'm not trying to vent at you was hoping to help another person out.  

Thanks for your patience on the matter I would've followed up and wasn't intending to ignore your thoughtful reply but I've just not had time since I posted my original post and reply to have time to tend to this.

The issue that I was running into was an Elastic Search one every time I work with the ElasticSearch container. I run into a wall. They must not have it designed to be fully stateless or something.  

Much respect for you making this project and I intended no disrespect.  If it would help improve the project I would gladly pick this up and see if I can file a PR or something to help improve the project.
